// e2b/dist/index.mjs
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __knownSymbol = (name, symbol) => (symbol = Symbol[name]) ? symbol : Symbol.for("Symbol." + name);
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};
var __await = function(promise, isYieldStar) {
  this[0] = promise;
  this[1] = isYieldStar;
};
var __asyncGenerator = (__this, __arguments, generator) => {
  var resume = (k, v, yes, no) => {
    try {
      var x = generator[k](v), isAwait = (v = x.value) instanceof __await, done = x.done;
      Promise.resolve(isAwait ? v[0] : v).then((y) => isAwait ? resume(k === "return" ? k : "next", v[1] ? { done: y.done, value: y.value } : y, yes, no) : yes({ value: y, done })).catch((e) => resume("throw", e, yes, no));
    } catch (e) {
      no(e);
    }
  }, method = (k) => it[k] = (x) => new Promise((yes, no) => resume(k, x, yes, no)), it = {};
  return generator = generator.apply(__this, __arguments), it[__knownSymbol("asyncIterator")] = () => it, method("next"), method("throw"), method("return"), it;
};
var __forAwait = (obj, it, method) => (it = obj[__knownSymbol("asyncIterator")]) ? it.call(obj) : (obj = obj[__knownSymbol("iterator")](), it = {}, method = (key, fn) => (fn = obj[key]) && (it[key] = (arg) => new Promise((yes, no, done) => (arg = fn.call(obj, arg), done = arg.done, Promise.resolve(arg.value).then((value) => yes({ value, done }), no)))), method("next"), method("return"), it);

// src/api/index.ts
import createClient from "openapi-fetch";

// src/api/metadata.ts
import platform from "platform";

// package.json
var version = "1.0.4";

// src/api/metadata.ts
function getRuntime() {
  var _a2, _b, _c;
  if (globalThis.Bun) {
    return { runtime: "bun", version: globalThis.Bun.version };
  }
  if (globalThis.Deno) {
    return { runtime: "deno", version: globalThis.Deno.version.deno };
  }
  if (((_b = (_a2 = globalThis.process) == null ? void 0 : _a2.release) == null ? void 0 : _b.name) === "node") {
    return { runtime: "node", version: platform.version || "unknown" };
  }
  if (typeof EdgeRuntime === "string") {
    return { runtime: "vercel-edge", version: "unknown" };
  }
  if (((_c = globalThis.navigator) == null ? void 0 : _c.userAgent) === "Cloudflare-Workers") {
    return { runtime: "cloudflare-worker", version: "unknown" };
  }
  if (typeof window !== "undefined") {
    return { runtime: "browser", version: platform.version || "unknown" };
  }
  return { runtime: "unknown", version: "unknown" };
}
var { runtime, version: runtimeVersion } = getRuntime();
var _a;
var defaultHeaders = {
  browser: typeof window !== "undefined" && platform.name || "unknown",
  lang: "js",
  lang_version: runtimeVersion,
  package_version: version,
  publisher: "e2b",
  sdk_runtime: runtime,
  system: ((_a = platform.os) == null ? void 0 : _a.family) || "unknown"
};
function getEnvVar(name) {
  if (runtime === "deno") {
    return Deno.env.get(name);
  }
  return process.env[name];
}

// src/errors.ts
function formatSandboxTimeoutError(message) {
  return new TimeoutError(
    `${message}: This error is likely due to sandbox timeout. You can modify the sandbox timeout by passing 'timeoutMs' when starting the sandbox or calling '.setTimeout' on the sandbox with the desired timeout.`
  );
}
var SandboxError = class extends Error {
  constructor(message) {
    super(message);
    this.name = "SandboxError";
  }
};
var TimeoutError = class extends SandboxError {
  constructor(message) {
    super(message);
    this.name = "TimeoutError";
  }
};
var InvalidArgumentError = class extends SandboxError {
  constructor(message) {
    super(message);
    this.name = "InvalidArgumentError";
  }
};
var NotEnoughSpaceError = class extends SandboxError {
  constructor(message) {
    super(message);
    this.name = "NotEnoughSpaceError";
  }
};
var NotFoundError = class extends SandboxError {
  constructor(message) {
    super(message);
    this.name = "NotFoundError";
  }
};
var AuthenticationError = class extends SandboxError {
  constructor(message) {
    super(message);
    this.name = "AuthenticationError";
  }
};
var TemplateError = class extends SandboxError {
  constructor(message) {
    super(message);
    this.name = "TemplateError";
  }
};
var RateLimitError = class extends SandboxError {
  constructor(message) {
    super(message);
    this.name = "RateLimitError";
  }
};

// src/logs.ts
function formatLog(log) {
  return JSON.parse(JSON.stringify(log));
}
function createRpcLogger(logger) {
  function logEach(stream) {
    return __asyncGenerator(this, null, function* () {
      var _a2;
      try {
        for (var iter = __forAwait(stream), more, temp, error; more = !(temp = yield new __await(iter.next())).done; more = false) {
          const m = temp.value;
          (_a2 = logger.debug) == null ? void 0 : _a2.call(logger, "Response stream:", formatLog(m));
          yield m;
        }
      } catch (temp) {
        error = [temp];
      } finally {
        try {
          more && (temp = iter.return) && (yield new __await(temp.call(iter)));
        } finally {
          if (error)
            throw error[0];
        }
      }
    });
  }
  return (next) => (req) => __async(this, null, function* () {
    var _a2, _b;
    (_a2 = logger.info) == null ? void 0 : _a2.call(logger, `Request: POST ${req.url}`);
    const res = yield next(req);
    if (res.stream) {
      return __spreadProps(__spreadValues({}, res), {
        message: logEach(res.message)
      });
    } else {
      (_b = logger.info) == null ? void 0 : _b.call(logger, "Response:", formatLog(res.message));
    }
    return res;
  });
}
function createApiLogger(logger) {
  return {
    onRequest(req) {
      return __async(this, null, function* () {
        var _a3;
        (_a3 = logger.info) == null ? void 0 : _a3.call(logger, `Request ${req.method} ${req.url}`);
        return req;
      });
    },
    onResponse(res) {
      return __async(this, null, function* () {
        var _a3, _b;
        if (res.status >= 400) {
          (_a3 = logger.error) == null ? void 0 : _a3.call(logger, "Response:", res.status, res.statusText);
        } else {
          (_b = logger.info) == null ? void 0 : _b.call(logger, "Response:", res.status, res.statusText);
        }
        return res;
      });
    }
  };
}

// src/api/index.ts
function handleApiError(response) {
  var _a2, _b;
  if (!response.error) {
    return;
  }
  if (response.response.status === 429) {
    return new RateLimitError("Rate limit exceeded, please try again later.");
  }
  const message = (_b = (_a2 = response.error) == null ? void 0 : _a2.message) != null ? _b : response.error;
  return new SandboxError(`${response.response.status}: ${message}`);
}
var ApiClient = class {
  constructor(config, opts = { requireAccessToken: false, requireApiKey: true }) {
    if (!(opts == null ? void 0 : opts.requireApiKey) && !config.apiKey) {
      throw new AuthenticationError(
        "API key is required, please visit https://e2b.dev/docs to get your API key. You can either set the environment variable `E2B_API_KEY` or you can pass it directly to the sandbox like Sandbox.create({ apiKey: 'e2b_...' })"
      );
    }
    if ((opts == null ? void 0 : opts.requireAccessToken) && !config.accessToken) {
      throw new AuthenticationError(
        "Access token is required, please visit https://e2b.dev/docs to get your access token. You can set the environment variable `E2B_ACCESS_TOKEN` or pass the `accessToken` in options."
      );
    }
    this.api = createClient({
      baseUrl: config.apiUrl,
      // keepalive: true, // TODO: Return keepalive
      headers: __spreadValues(__spreadValues(__spreadValues({}, defaultHeaders), config.apiKey && { "X-API-KEY": config.apiKey }), config.accessToken && {
        Authorization: `Bearer ${config.accessToken}`
      })
    });
    if (config.logger) {
      this.api.use(createApiLogger(config.logger));
    }
  }
};

// src/connectionConfig.ts
var REQUEST_TIMEOUT_MS = 3e4;
var KEEPALIVE_PING_INTERVAL_SEC = 50;
var KEEPALIVE_PING_HEADER = "Keepalive-Ping-Interval";
var ConnectionConfig = class _ConnectionConfig {
  constructor(opts) {
    var _a2;
    this.apiKey = (opts == null ? void 0 : opts.apiKey) || _ConnectionConfig.apiKey;
    this.debug = (opts == null ? void 0 : opts.debug) || _ConnectionConfig.debug;
    this.domain = (opts == null ? void 0 : opts.domain) || _ConnectionConfig.domain;
    this.accessToken = (opts == null ? void 0 : opts.accessToken) || _ConnectionConfig.accessToken;
    this.requestTimeoutMs = (_a2 = opts == null ? void 0 : opts.requestTimeoutMs) != null ? _a2 : REQUEST_TIMEOUT_MS;
    this.logger = opts == null ? void 0 : opts.logger;
    this.apiUrl = this.debug ? "http://localhost:3000" : `https://api.${this.domain}`;
  }
  static get domain() {
    return getEnvVar("E2B_DOMAIN") || "e2b.dev";
  }
  static get debug() {
    return (getEnvVar("E2B_DEBUG") || "false").toLowerCase() === "true";
  }
  static get apiKey() {
    return getEnvVar("E2B_API_KEY");
  }
  static get accessToken() {
    return getEnvVar("E2B_ACCESS_TOKEN");
  }
  getSignal(requestTimeoutMs) {
    const timeout = requestTimeoutMs != null ? requestTimeoutMs : this.requestTimeoutMs;
    return timeout ? AbortSignal.timeout(timeout) : void 0;
  }
};
var defaultUsername = "user";

// src/envd/rpc.ts
import { Code, ConnectError } from "@connectrpc/connect";
function handleRpcError(err) {
  if (err instanceof ConnectError) {
    switch (err.code) {
      case Code.InvalidArgument:
        return new InvalidArgumentError(err.message);
      case Code.Unauthenticated:
        return new AuthenticationError(err.message);
      case Code.NotFound:
        return new NotFoundError(err.message);
      case Code.Unavailable:
        return formatSandboxTimeoutError(err.message);
      case Code.Canceled:
        return new TimeoutError(
          `${err.message}: This error is likely due to exceeding 'requestTimeoutMs'. You can pass the request timeout value as an option when making the request.`
        );
      case Code.DeadlineExceeded:
        return new TimeoutError(
          `${err.message}: This error is likely due to exceeding 'timeoutMs' \u2014 the total time a long running request (like command execution or directory watch) can be active. It can be modified by passing 'timeoutMs' when making the request. Use '0' to disable the timeout.`
        );
      default:
        return new SandboxError(`${err.code}: ${err.message}`);
    }
  }
  return err;
}
function encode64(value) {
  switch (runtime) {
    case "deno":
      return btoa(value);
    case "node":
      return Buffer.from(value).toString("base64");
    case "bun":
      return Buffer.from(value).toString("base64");
    default:
      return btoa(value);
  }
}
function authenticationHeader(username) {
  const value = `${username || defaultUsername}:`;
  const encoded = encode64(value);
  return { "Authorization": `Basic ${encoded}` };
}

// src/envd/filesystem/filesystem_pb.ts
import { Message, proto3 } from "@bufbuild/protobuf";
var FileType = /* @__PURE__ */ ((FileType3) => {
  FileType3[FileType3["UNSPECIFIED"] = 0] = "UNSPECIFIED";
  FileType3[FileType3["FILE"] = 1] = "FILE";
  FileType3[FileType3["DIRECTORY"] = 2] = "DIRECTORY";
  return FileType3;
})(FileType || {});
proto3.util.setEnumType(FileType, "filesystem.FileType", [
  { no: 0, name: "FILE_TYPE_UNSPECIFIED" },
  { no: 1, name: "FILE_TYPE_FILE" },
  { no: 2, name: "FILE_TYPE_DIRECTORY" }
]);
var EventType = /* @__PURE__ */ ((EventType2) => {
  EventType2[EventType2["UNSPECIFIED"] = 0] = "UNSPECIFIED";
  EventType2[EventType2["CREATE"] = 1] = "CREATE";
  EventType2[EventType2["WRITE"] = 2] = "WRITE";
  EventType2[EventType2["REMOVE"] = 3] = "REMOVE";
  EventType2[EventType2["RENAME"] = 4] = "RENAME";
  EventType2[EventType2["CHMOD"] = 5] = "CHMOD";
  return EventType2;
})(EventType || {});
proto3.util.setEnumType(EventType, "filesystem.EventType", [
  { no: 0, name: "EVENT_TYPE_UNSPECIFIED" },
  { no: 1, name: "EVENT_TYPE_CREATE" },
  { no: 2, name: "EVENT_TYPE_WRITE" },
  { no: 3, name: "EVENT_TYPE_REMOVE" },
  { no: 4, name: "EVENT_TYPE_RENAME" },
  { no: 5, name: "EVENT_TYPE_CHMOD" }
]);
var _MoveRequest = class _MoveRequest extends Message {
  constructor(data) {
    super();
    /**
     * @generated from field: string source = 1;
     */
    this.source = "";
    /**
     * @generated from field: string destination = 2;
     */
    this.destination = "";
    proto3.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _MoveRequest().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _MoveRequest().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _MoveRequest().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto3.util.equals(_MoveRequest, a, b);
  }
};
_MoveRequest.runtime = proto3;
_MoveRequest.typeName = "filesystem.MoveRequest";
_MoveRequest.fields = proto3.util.newFieldList(() => [
  {
    no: 1,
    name: "source",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  {
    no: 2,
    name: "destination",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]);
var MoveRequest = _MoveRequest;
var _MoveResponse = class _MoveResponse extends Message {
  constructor(data) {
    super();
    proto3.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _MoveResponse().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _MoveResponse().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _MoveResponse().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto3.util.equals(_MoveResponse, a, b);
  }
};
_MoveResponse.runtime = proto3;
_MoveResponse.typeName = "filesystem.MoveResponse";
_MoveResponse.fields = proto3.util.newFieldList(() => [
  { no: 1, name: "entry", kind: "message", T: EntryInfo }
]);
var MoveResponse = _MoveResponse;
var _MakeDirRequest = class _MakeDirRequest extends Message {
  constructor(data) {
    super();
    /**
     * @generated from field: string path = 1;
     */
    this.path = "";
    proto3.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _MakeDirRequest().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _MakeDirRequest().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _MakeDirRequest().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto3.util.equals(_MakeDirRequest, a, b);
  }
};
_MakeDirRequest.runtime = proto3;
_MakeDirRequest.typeName = "filesystem.MakeDirRequest";
_MakeDirRequest.fields = proto3.util.newFieldList(() => [
  {
    no: 1,
    name: "path",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]);
var MakeDirRequest = _MakeDirRequest;
var _MakeDirResponse = class _MakeDirResponse extends Message {
  constructor(data) {
    super();
    proto3.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _MakeDirResponse().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _MakeDirResponse().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _MakeDirResponse().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto3.util.equals(_MakeDirResponse, a, b);
  }
};
_MakeDirResponse.runtime = proto3;
_MakeDirResponse.typeName = "filesystem.MakeDirResponse";
_MakeDirResponse.fields = proto3.util.newFieldList(() => [
  { no: 1, name: "entry", kind: "message", T: EntryInfo }
]);
var MakeDirResponse = _MakeDirResponse;
var _RemoveRequest = class _RemoveRequest extends Message {
  constructor(data) {
    super();
    /**
     * @generated from field: string path = 1;
     */
    this.path = "";
    proto3.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _RemoveRequest().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _RemoveRequest().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _RemoveRequest().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto3.util.equals(_RemoveRequest, a, b);
  }
};
_RemoveRequest.runtime = proto3;
_RemoveRequest.typeName = "filesystem.RemoveRequest";
_RemoveRequest.fields = proto3.util.newFieldList(() => [
  {
    no: 1,
    name: "path",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]);
var RemoveRequest = _RemoveRequest;
var _RemoveResponse = class _RemoveResponse extends Message {
  constructor(data) {
    super();
    proto3.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _RemoveResponse().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _RemoveResponse().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _RemoveResponse().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto3.util.equals(_RemoveResponse, a, b);
  }
};
_RemoveResponse.runtime = proto3;
_RemoveResponse.typeName = "filesystem.RemoveResponse";
_RemoveResponse.fields = proto3.util.newFieldList(() => []);
var RemoveResponse = _RemoveResponse;
var _StatRequest = class _StatRequest extends Message {
  constructor(data) {
    super();
    /**
     * @generated from field: string path = 1;
     */
    this.path = "";
    proto3.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _StatRequest().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _StatRequest().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _StatRequest().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto3.util.equals(_StatRequest, a, b);
  }
};
_StatRequest.runtime = proto3;
_StatRequest.typeName = "filesystem.StatRequest";
_StatRequest.fields = proto3.util.newFieldList(() => [
  {
    no: 1,
    name: "path",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]);
var StatRequest = _StatRequest;
var _StatResponse = class _StatResponse extends Message {
  constructor(data) {
    super();
    proto3.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _StatResponse().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _StatResponse().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _StatResponse().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto3.util.equals(_StatResponse, a, b);
  }
};
_StatResponse.runtime = proto3;
_StatResponse.typeName = "filesystem.StatResponse";
_StatResponse.fields = proto3.util.newFieldList(() => [
  { no: 1, name: "entry", kind: "message", T: EntryInfo }
]);
var StatResponse = _StatResponse;
var _EntryInfo = class _EntryInfo extends Message {
  constructor(data) {
    super();
    /**
     * @generated from field: string name = 1;
     */
    this.name = "";
    /**
     * @generated from field: filesystem.FileType type = 2;
     */
    this.type = 0 /* UNSPECIFIED */;
    /**
     * @generated from field: string path = 3;
     */
    this.path = "";
    proto3.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _EntryInfo().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _EntryInfo().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _EntryInfo().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto3.util.equals(_EntryInfo, a, b);
  }
};
_EntryInfo.runtime = proto3;
_EntryInfo.typeName = "filesystem.EntryInfo";
_EntryInfo.fields = proto3.util.newFieldList(() => [
  {
    no: 1,
    name: "name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  { no: 2, name: "type", kind: "enum", T: proto3.getEnumType(FileType) },
  {
    no: 3,
    name: "path",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]);
var EntryInfo = _EntryInfo;
var _ListDirRequest = class _ListDirRequest extends Message {
  constructor(data) {
    super();
    /**
     * @generated from field: string path = 1;
     */
    this.path = "";
    proto3.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _ListDirRequest().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _ListDirRequest().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _ListDirRequest().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto3.util.equals(_ListDirRequest, a, b);
  }
};
_ListDirRequest.runtime = proto3;
_ListDirRequest.typeName = "filesystem.ListDirRequest";
_ListDirRequest.fields = proto3.util.newFieldList(() => [
  {
    no: 1,
    name: "path",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]);
var ListDirRequest = _ListDirRequest;
var _ListDirResponse = class _ListDirResponse extends Message {
  constructor(data) {
    super();
    /**
     * @generated from field: repeated filesystem.EntryInfo entries = 1;
     */
    this.entries = [];
    proto3.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _ListDirResponse().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _ListDirResponse().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _ListDirResponse().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto3.util.equals(_ListDirResponse, a, b);
  }
};
_ListDirResponse.runtime = proto3;
_ListDirResponse.typeName = "filesystem.ListDirResponse";
_ListDirResponse.fields = proto3.util.newFieldList(() => [
  { no: 1, name: "entries", kind: "message", T: EntryInfo, repeated: true }
]);
var ListDirResponse = _ListDirResponse;
var _WatchDirRequest = class _WatchDirRequest extends Message {
  constructor(data) {
    super();
    /**
     * @generated from field: string path = 1;
     */
    this.path = "";
    proto3.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _WatchDirRequest().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _WatchDirRequest().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _WatchDirRequest().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto3.util.equals(_WatchDirRequest, a, b);
  }
};
_WatchDirRequest.runtime = proto3;
_WatchDirRequest.typeName = "filesystem.WatchDirRequest";
_WatchDirRequest.fields = proto3.util.newFieldList(() => [
  {
    no: 1,
    name: "path",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]);
var WatchDirRequest = _WatchDirRequest;
var _FilesystemEvent = class _FilesystemEvent extends Message {
  constructor(data) {
    super();
    /**
     * @generated from field: string name = 1;
     */
    this.name = "";
    /**
     * @generated from field: filesystem.EventType type = 2;
     */
    this.type = 0 /* UNSPECIFIED */;
    proto3.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _FilesystemEvent().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _FilesystemEvent().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _FilesystemEvent().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto3.util.equals(_FilesystemEvent, a, b);
  }
};
_FilesystemEvent.runtime = proto3;
_FilesystemEvent.typeName = "filesystem.FilesystemEvent";
_FilesystemEvent.fields = proto3.util.newFieldList(() => [
  {
    no: 1,
    name: "name",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  { no: 2, name: "type", kind: "enum", T: proto3.getEnumType(EventType) }
]);
var FilesystemEvent = _FilesystemEvent;
var _WatchDirResponse = class _WatchDirResponse extends Message {
  constructor(data) {
    super();
    /**
     * @generated from oneof filesystem.WatchDirResponse.event
     */
    this.event = { case: void 0 };
    proto3.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _WatchDirResponse().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _WatchDirResponse().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _WatchDirResponse().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto3.util.equals(_WatchDirResponse, a, b);
  }
};
_WatchDirResponse.runtime = proto3;
_WatchDirResponse.typeName = "filesystem.WatchDirResponse";
_WatchDirResponse.fields = proto3.util.newFieldList(() => [
  { no: 1, name: "start", kind: "message", T: WatchDirResponse_StartEvent, oneof: "event" },
  { no: 2, name: "filesystem", kind: "message", T: FilesystemEvent, oneof: "event" },
  { no: 3, name: "keepalive", kind: "message", T: WatchDirResponse_KeepAlive, oneof: "event" }
]);
var WatchDirResponse = _WatchDirResponse;
var _WatchDirResponse_StartEvent = class _WatchDirResponse_StartEvent extends Message {
  constructor(data) {
    super();
    proto3.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _WatchDirResponse_StartEvent().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _WatchDirResponse_StartEvent().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _WatchDirResponse_StartEvent().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto3.util.equals(_WatchDirResponse_StartEvent, a, b);
  }
};
_WatchDirResponse_StartEvent.runtime = proto3;
_WatchDirResponse_StartEvent.typeName = "filesystem.WatchDirResponse.StartEvent";
_WatchDirResponse_StartEvent.fields = proto3.util.newFieldList(() => []);
var WatchDirResponse_StartEvent = _WatchDirResponse_StartEvent;
var _WatchDirResponse_KeepAlive = class _WatchDirResponse_KeepAlive extends Message {
  constructor(data) {
    super();
    proto3.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _WatchDirResponse_KeepAlive().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _WatchDirResponse_KeepAlive().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _WatchDirResponse_KeepAlive().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto3.util.equals(_WatchDirResponse_KeepAlive, a, b);
  }
};
_WatchDirResponse_KeepAlive.runtime = proto3;
_WatchDirResponse_KeepAlive.typeName = "filesystem.WatchDirResponse.KeepAlive";
_WatchDirResponse_KeepAlive.fields = proto3.util.newFieldList(() => []);
var WatchDirResponse_KeepAlive = _WatchDirResponse_KeepAlive;
var _CreateWatcherRequest = class _CreateWatcherRequest extends Message {
  constructor(data) {
    super();
    /**
     * @generated from field: string path = 1;
     */
    this.path = "";
    proto3.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _CreateWatcherRequest().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _CreateWatcherRequest().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _CreateWatcherRequest().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto3.util.equals(_CreateWatcherRequest, a, b);
  }
};
_CreateWatcherRequest.runtime = proto3;
_CreateWatcherRequest.typeName = "filesystem.CreateWatcherRequest";
_CreateWatcherRequest.fields = proto3.util.newFieldList(() => [
  {
    no: 1,
    name: "path",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]);
var CreateWatcherRequest = _CreateWatcherRequest;
var _CreateWatcherResponse = class _CreateWatcherResponse extends Message {
  constructor(data) {
    super();
    /**
     * @generated from field: string watcher_id = 1;
     */
    this.watcherId = "";
    proto3.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _CreateWatcherResponse().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _CreateWatcherResponse().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _CreateWatcherResponse().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto3.util.equals(_CreateWatcherResponse, a, b);
  }
};
_CreateWatcherResponse.runtime = proto3;
_CreateWatcherResponse.typeName = "filesystem.CreateWatcherResponse";
_CreateWatcherResponse.fields = proto3.util.newFieldList(() => [
  {
    no: 1,
    name: "watcher_id",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]);
var CreateWatcherResponse = _CreateWatcherResponse;
var _GetWatcherEventsRequest = class _GetWatcherEventsRequest extends Message {
  constructor(data) {
    super();
    /**
     * @generated from field: string watcher_id = 1;
     */
    this.watcherId = "";
    proto3.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _GetWatcherEventsRequest().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _GetWatcherEventsRequest().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _GetWatcherEventsRequest().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto3.util.equals(_GetWatcherEventsRequest, a, b);
  }
};
_GetWatcherEventsRequest.runtime = proto3;
_GetWatcherEventsRequest.typeName = "filesystem.GetWatcherEventsRequest";
_GetWatcherEventsRequest.fields = proto3.util.newFieldList(() => [
  {
    no: 1,
    name: "watcher_id",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]);
var GetWatcherEventsRequest = _GetWatcherEventsRequest;
var _GetWatcherEventsResponse = class _GetWatcherEventsResponse extends Message {
  constructor(data) {
    super();
    /**
     * @generated from field: repeated filesystem.FilesystemEvent events = 1;
     */
    this.events = [];
    proto3.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _GetWatcherEventsResponse().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _GetWatcherEventsResponse().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _GetWatcherEventsResponse().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto3.util.equals(_GetWatcherEventsResponse, a, b);
  }
};
_GetWatcherEventsResponse.runtime = proto3;
_GetWatcherEventsResponse.typeName = "filesystem.GetWatcherEventsResponse";
_GetWatcherEventsResponse.fields = proto3.util.newFieldList(() => [
  { no: 1, name: "events", kind: "message", T: FilesystemEvent, repeated: true }
]);
var GetWatcherEventsResponse = _GetWatcherEventsResponse;
var _RemoveWatcherRequest = class _RemoveWatcherRequest extends Message {
  constructor(data) {
    super();
    /**
     * @generated from field: string watcher_id = 1;
     */
    this.watcherId = "";
    proto3.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _RemoveWatcherRequest().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _RemoveWatcherRequest().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _RemoveWatcherRequest().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto3.util.equals(_RemoveWatcherRequest, a, b);
  }
};
_RemoveWatcherRequest.runtime = proto3;
_RemoveWatcherRequest.typeName = "filesystem.RemoveWatcherRequest";
_RemoveWatcherRequest.fields = proto3.util.newFieldList(() => [
  {
    no: 1,
    name: "watcher_id",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  }
]);
var RemoveWatcherRequest = _RemoveWatcherRequest;
var _RemoveWatcherResponse = class _RemoveWatcherResponse extends Message {
  constructor(data) {
    super();
    proto3.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _RemoveWatcherResponse().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _RemoveWatcherResponse().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _RemoveWatcherResponse().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto3.util.equals(_RemoveWatcherResponse, a, b);
  }
};
_RemoveWatcherResponse.runtime = proto3;
_RemoveWatcherResponse.typeName = "filesystem.RemoveWatcherResponse";
_RemoveWatcherResponse.fields = proto3.util.newFieldList(() => []);
var RemoveWatcherResponse = _RemoveWatcherResponse;

// src/sandbox/filesystem/watchHandle.ts
var FilesystemEventType = /* @__PURE__ */ ((FilesystemEventType2) => {
  FilesystemEventType2["CHMOD"] = "chmod";
  FilesystemEventType2["CREATE"] = "create";
  FilesystemEventType2["REMOVE"] = "remove";
  FilesystemEventType2["RENAME"] = "rename";
  FilesystemEventType2["WRITE"] = "write";
  return FilesystemEventType2;
})(FilesystemEventType || {});
function mapEventType(type) {
  switch (type) {
    case 5 /* CHMOD */:
      return "chmod" /* CHMOD */;
    case 1 /* CREATE */:
      return "create" /* CREATE */;
    case 3 /* REMOVE */:
      return "remove" /* REMOVE */;
    case 4 /* RENAME */:
      return "rename" /* RENAME */;
    case 2 /* WRITE */:
      return "write" /* WRITE */;
  }
}
var WatchHandle = class {
  constructor(handleStop, events, onEvent, onExit) {
    this.handleStop = handleStop;
    this.events = events;
    this.onEvent = onEvent;
    this.onExit = onExit;
    this.handleEvents();
  }
  /**
   * Stop watching the directory.
   */
  stop() {
    return __async(this, null, function* () {
      this.handleStop();
    });
  }
  iterateEvents() {
    return __asyncGenerator(this, null, function* () {
      try {
        try {
          for (var iter = __forAwait(this.events), more, temp, error; more = !(temp = yield new __await(iter.next())).done; more = false) {
            const event = temp.value;
            switch (event.event.case) {
              case "filesystem":
                yield event.event;
                break;
            }
          }
        } catch (temp) {
          error = [temp];
        } finally {
          try {
            more && (temp = iter.return) && (yield new __await(temp.call(iter)));
          } finally {
            if (error)
              throw error[0];
          }
        }
      } catch (err) {
        throw handleRpcError(err);
      }
    });
  }
  handleEvents() {
    return __async(this, null, function* () {
      var _a2, _b, _c;
      try {
        try {
          for (var iter = __forAwait(this.iterateEvents()), more, temp, error; more = !(temp = yield iter.next()).done; more = false) {
            const event = temp.value;
            const eventType = mapEventType(event.value.type);
            if (eventType === void 0) {
              continue;
            }
            (_a2 = this.onEvent) == null ? void 0 : _a2.call(this, {
              name: event.value.name,
              type: eventType
            });
          }
        } catch (temp) {
          error = [temp];
        } finally {
          try {
            more && (temp = iter.return) && (yield temp.call(iter));
          } finally {
            if (error)
              throw error[0];
          }
        }
        (_b = this.onExit) == null ? void 0 : _b.call(this);
      } catch (err) {
        (_c = this.onExit) == null ? void 0 : _c.call(this, err);
      }
    });
  }
};

// src/sandbox/filesystem/index.ts
import {
  createPromiseClient,
  ConnectError as ConnectError3,
  Code as Code3
} from "@connectrpc/connect";

// src/envd/api.ts
import createClient2 from "openapi-fetch";
import { Code as Code2, ConnectError as ConnectError2 } from "@connectrpc/connect";
function handleEnvdApiError(res) {
  return __async(this, null, function* () {
    var _a2;
    if (!res.error) {
      return;
    }
    const message = typeof res.error == "string" ? res.error : ((_a2 = res.error) == null ? void 0 : _a2.message) || (yield res.response.text());
    switch (res.response.status) {
      case 400:
        return new InvalidArgumentError(message);
      case 401:
        return new AuthenticationError(message);
      case 404:
        return new NotFoundError(message);
      case 429:
        return new SandboxError(
          `${res.response.status}: ${message}: The requests are being rate limited.`
        );
      case 502:
        return formatSandboxTimeoutError(message);
      case 507:
        return new NotEnoughSpaceError(message);
      default:
        return new SandboxError(`${res.response.status}: ${message}`);
    }
  });
}
function handleProcessStartEvent(events) {
  return __async(this, null, function* () {
    var _a2;
    let startEvent;
    try {
      startEvent = (yield events[Symbol.asyncIterator]().next()).value;
    } catch (err) {
      if (err instanceof ConnectError2) {
        if (err.code === Code2.Unavailable) {
          throw new NotFoundError("Sandbox is probably not running anymore");
        }
      }
      throw err;
    }
    if (((_a2 = startEvent.event) == null ? void 0 : _a2.event.case) !== "start") {
      throw new Error("Expected start event");
    }
    return startEvent.event.event.value.pid;
  });
}
function handleWatchDirStartEvent(events) {
  return __async(this, null, function* () {
    var _a2;
    let startEvent;
    try {
      startEvent = (yield events[Symbol.asyncIterator]().next()).value;
    } catch (err) {
      if (err instanceof ConnectError2) {
        if (err.code === Code2.Unavailable) {
          throw new NotFoundError("Sandbox is probably not running anymore");
        }
      }
      throw err;
    }
    if (((_a2 = startEvent.event) == null ? void 0 : _a2.case) !== "start") {
      throw new Error("Expected start event");
    }
    return startEvent.event.value;
  });
}
var EnvdApiClient = class {
  constructor(config) {
    this.api = createClient2({
      baseUrl: config.apiUrl
      // keepalive: true, // TODO: Return keepalive
    });
    if (config.logger) {
      this.api.use(createApiLogger(config.logger));
    }
  }
};

// src/envd/filesystem/filesystem_connect.ts
import { MethodKind } from "@bufbuild/protobuf";
var Filesystem = {
  typeName: "filesystem.Filesystem",
  methods: {
    /**
     * @generated from rpc filesystem.Filesystem.Stat
     */
    stat: {
      name: "Stat",
      I: StatRequest,
      O: StatResponse,
      kind: MethodKind.Unary
    },
    /**
     * @generated from rpc filesystem.Filesystem.MakeDir
     */
    makeDir: {
      name: "MakeDir",
      I: MakeDirRequest,
      O: MakeDirResponse,
      kind: MethodKind.Unary
    },
    /**
     * @generated from rpc filesystem.Filesystem.Move
     */
    move: {
      name: "Move",
      I: MoveRequest,
      O: MoveResponse,
      kind: MethodKind.Unary
    },
    /**
     * @generated from rpc filesystem.Filesystem.ListDir
     */
    listDir: {
      name: "ListDir",
      I: ListDirRequest,
      O: ListDirResponse,
      kind: MethodKind.Unary
    },
    /**
     * @generated from rpc filesystem.Filesystem.Remove
     */
    remove: {
      name: "Remove",
      I: RemoveRequest,
      O: RemoveResponse,
      kind: MethodKind.Unary
    },
    /**
     * @generated from rpc filesystem.Filesystem.WatchDir
     */
    watchDir: {
      name: "WatchDir",
      I: WatchDirRequest,
      O: WatchDirResponse,
      kind: MethodKind.ServerStreaming
    },
    /**
     * Non-streaming versions of WatchDir
     *
     * @generated from rpc filesystem.Filesystem.CreateWatcher
     */
    createWatcher: {
      name: "CreateWatcher",
      I: CreateWatcherRequest,
      O: CreateWatcherResponse,
      kind: MethodKind.Unary
    },
    /**
     * @generated from rpc filesystem.Filesystem.GetWatcherEvents
     */
    getWatcherEvents: {
      name: "GetWatcherEvents",
      I: GetWatcherEventsRequest,
      O: GetWatcherEventsResponse,
      kind: MethodKind.Unary
    },
    /**
     * @generated from rpc filesystem.Filesystem.RemoveWatcher
     */
    removeWatcher: {
      name: "RemoveWatcher",
      I: RemoveWatcherRequest,
      O: RemoveWatcherResponse,
      kind: MethodKind.Unary
    }
  }
};

// src/sandbox/filesystem/index.ts
var FileType2 = /* @__PURE__ */ ((FileType3) => {
  FileType3["FILE"] = "file";
  FileType3["DIR"] = "dir";
  return FileType3;
})(FileType2 || {});
function mapFileType(fileType) {
  switch (fileType) {
    case 2 /* DIRECTORY */:
      return "dir" /* DIR */;
    case 1 /* FILE */:
      return "file" /* FILE */;
  }
}
var Filesystem2 = class {
  // 60 seconds
  constructor(transport, envdApi, connectionConfig) {
    this.envdApi = envdApi;
    this.connectionConfig = connectionConfig;
    this.defaultWatchTimeout = 6e4;
    this.rpc = createPromiseClient(Filesystem, transport);
  }
  read(path, opts) {
    return __async(this, null, function* () {
      var _a2;
      const format = (_a2 = opts == null ? void 0 : opts.format) != null ? _a2 : "text";
      const res = yield this.envdApi.api.GET("/files", {
        params: {
          query: {
            path,
            username: (opts == null ? void 0 : opts.user) || defaultUsername
          }
        },
        parseAs: format === "bytes" ? "arrayBuffer" : format,
        signal: this.connectionConfig.getSignal(opts == null ? void 0 : opts.requestTimeoutMs)
      });
      const err = yield handleEnvdApiError(res);
      if (err) {
        throw err;
      }
      if (format === "bytes") {
        return new Uint8Array(res.data);
      }
      if (res.response.headers.get("content-length") === "0") {
        return "";
      }
      return res.data;
    });
  }
  /**
   * Write content to a file.
   *
   *
   * Writing to a file that doesn't exist creates the file.
   *
   * Writing to a file that already exists overwrites the file.
   *
   * Writing to a file at path that doesn't exist creates the necessary directories.
   *
   * @param path path to file.
   * @param data data to write to the file. Data can be a string, `ArrayBuffer`, `Blob`, or `ReadableStream`.
   * @param opts connection options.
   *
   * @returns information about the written file
   */
  write(path, data, opts) {
    return __async(this, null, function* () {
      const blob = yield new Response(data).blob();
      const res = yield this.envdApi.api.POST("/files", {
        params: {
          query: {
            path,
            username: (opts == null ? void 0 : opts.user) || defaultUsername
          }
        },
        bodySerializer() {
          const fd = new FormData();
          fd.append("file", blob);
          return fd;
        },
        body: {},
        headers: {
          "Content-Type": "multipart/form-data",
          "Bun-Content-Type": "temporary-fix"
          // https://github.com/oven-sh/bun/issues/14988
        }
      });
      const err = yield handleEnvdApiError(res);
      if (err) {
        throw err;
      }
      const files = res.data;
      if (!files || files.length === 0) {
        throw new Error("Expected to receive information about written file");
      }
      return files[0];
    });
  }
  /**
   * List entries in a directory.
   *
   * @param path path to the directory.
   * @param opts connection options.
   *
   * @returns list of entries in the sandbox filesystem directory.
   */
  list(path, opts) {
    return __async(this, null, function* () {
      try {
        const res = yield this.rpc.listDir(
          { path },
          {
            headers: authenticationHeader(opts == null ? void 0 : opts.user),
            signal: this.connectionConfig.getSignal(opts == null ? void 0 : opts.requestTimeoutMs)
          }
        );
        const entries = [];
        for (const e of res.entries) {
          const type = mapFileType(e.type);
          if (type) {
            entries.push({
              name: e.name,
              type,
              path: e.path
            });
          }
        }
        return entries;
      } catch (err) {
        throw handleRpcError(err);
      }
    });
  }
  /**
   * Create a new directory and all directories along the way if needed on the specified path.
   *
   * @param path path to a new directory. For example '/dirA/dirB' when creating 'dirB'.
   * @param opts connection options.
   *
   * @returns `true` if the directory was created, `false` if it already exists.
   */
  makeDir(path, opts) {
    return __async(this, null, function* () {
      try {
        yield this.rpc.makeDir(
          { path },
          {
            headers: authenticationHeader(opts == null ? void 0 : opts.user),
            signal: this.connectionConfig.getSignal(opts == null ? void 0 : opts.requestTimeoutMs)
          }
        );
        return true;
      } catch (err) {
        if (err instanceof ConnectError3) {
          if (err.code === Code3.AlreadyExists) {
            return false;
          }
        }
        throw handleRpcError(err);
      }
    });
  }
  /**
   * Rename a file or directory.
   *
   * @param oldPath path to the file or directory to rename.
   * @param newPath new path for the file or directory.
   * @param opts connection options.
   *
   * @returns information about renamed file or directory.
   */
  rename(oldPath, newPath, opts) {
    return __async(this, null, function* () {
      try {
        const res = yield this.rpc.move(
          {
            source: oldPath,
            destination: newPath
          },
          {
            headers: authenticationHeader(opts == null ? void 0 : opts.user),
            signal: this.connectionConfig.getSignal(opts == null ? void 0 : opts.requestTimeoutMs)
          }
        );
        const entry = res.entry;
        if (!entry) {
          throw new Error("Expected to receive information about moved object");
        }
        return {
          name: entry.name,
          type: mapFileType(entry.type),
          path: entry.path
        };
      } catch (err) {
        throw handleRpcError(err);
      }
    });
  }
  /**
   * Remove a file or directory.
   *
   * @param path path to a file or directory.
   * @param opts connection options.
   */
  remove(path, opts) {
    return __async(this, null, function* () {
      try {
        yield this.rpc.remove(
          { path },
          {
            headers: authenticationHeader(opts == null ? void 0 : opts.user),
            signal: this.connectionConfig.getSignal(opts == null ? void 0 : opts.requestTimeoutMs)
          }
        );
      } catch (err) {
        throw handleRpcError(err);
      }
    });
  }
  /**
   * Check if a file or a directory exists.
   *
   * @param path path to a file or a directory
   * @param opts connection options.
   *
   * @returns `true` if the file or directory exists, `false` otherwise
   */
  exists(path, opts) {
    return __async(this, null, function* () {
      try {
        yield this.rpc.stat(
          { path },
          {
            headers: authenticationHeader(opts == null ? void 0 : opts.user),
            signal: this.connectionConfig.getSignal(opts == null ? void 0 : opts.requestTimeoutMs)
          }
        );
        return true;
      } catch (err) {
        if (err instanceof ConnectError3) {
          if (err.code === Code3.NotFound) {
            return false;
          }
        }
        throw handleRpcError(err);
      }
    });
  }
  /**
   * Start watching a directory for filesystem events.
   *
   * @param path path to directory to watch.
   * @param onEvent callback to call when an event in the directory occurs.
   * @param opts connection options.
   *
   * @returns `WatchHandle` object for stopping watching directory.
   */
  watchDir(path, onEvent, opts) {
    return __async(this, null, function* () {
      var _a2, _b;
      const requestTimeoutMs = (_a2 = opts == null ? void 0 : opts.requestTimeoutMs) != null ? _a2 : this.connectionConfig.requestTimeoutMs;
      const controller = new AbortController();
      const reqTimeout = requestTimeoutMs ? setTimeout(() => {
        controller.abort();
      }, requestTimeoutMs) : void 0;
      const events = this.rpc.watchDir(
        { path },
        {
          headers: __spreadProps(__spreadValues({}, authenticationHeader(opts == null ? void 0 : opts.user)), {
            [KEEPALIVE_PING_HEADER]: KEEPALIVE_PING_INTERVAL_SEC.toString()
          }),
          signal: controller.signal,
          timeoutMs: (_b = opts == null ? void 0 : opts.timeoutMs) != null ? _b : this.defaultWatchTimeout
        }
      );
      try {
        yield handleWatchDirStartEvent(events);
        clearTimeout(reqTimeout);
        return new WatchHandle(
          () => controller.abort(),
          events,
          onEvent,
          opts == null ? void 0 : opts.onExit
        );
      } catch (err) {
        throw handleRpcError(err);
      }
    });
  }
};

// src/sandbox/commands/commandHandle.ts
var CommandExitError = class extends SandboxError {
  constructor(result) {
    super(result.error);
    this.result = result;
    this.name = "CommandExitError";
  }
  /**
   * Command execution exit code.
   * `0` if the command finished successfully.
   */
  get exitCode() {
    return this.result.exitCode;
  }
  /**
   * Error message from command execution.
   */
  get error() {
    return this.result.error;
  }
  /**
   * Command execution stdout output.
   */
  get stdout() {
    return this.result.stdout;
  }
  /**
   * Command execution stderr output.
   */
  get stderr() {
    return this.result.stderr;
  }
};
var CommandHandle = class {
  /**
   * @hidden
   * @internal
   * @access protected
   */
  constructor(pid, handleDisconnect, handleKill, events, onStdout, onStderr, onPty) {
    this.pid = pid;
    this.handleDisconnect = handleDisconnect;
    this.handleKill = handleKill;
    this.events = events;
    this.onStdout = onStdout;
    this.onStderr = onStderr;
    this.onPty = onPty;
    this._stdout = "";
    this._stderr = "";
    this._wait = this.handleEvents();
  }
  /**
   * Command execution exit code.
   * `0` if the command finished successfully.
   * 
   * It is `undefined` if the command is still running.
   */
  get exitCode() {
    var _a2;
    return (_a2 = this.result) == null ? void 0 : _a2.exitCode;
  }
  /**
   * Error message from command execution.
   */
  get error() {
    var _a2;
    return (_a2 = this.result) == null ? void 0 : _a2.error;
  }
  /**
   * Command execution stderr output.
   */
  get stderr() {
    return this._stderr;
  }
  /**
   * Command execution stdout output.
   */
  get stdout() {
    return this._stdout;
  }
  /**
   * Wait for the command to finish and return the result.
   * If the command exits with a non-zero exit code, it throws a `CommandExitError`.
   * 
   * @returns `CommandResult` result of command execution.
   */
  wait() {
    return __async(this, null, function* () {
      yield this._wait;
      if (this.iterationError) {
        throw this.iterationError;
      }
      if (!this.result) {
        throw new SandboxError("Process exited without a result");
      }
      if (this.result.exitCode !== 0) {
        throw new CommandExitError(this.result);
      }
      return this.result;
    });
  }
  /**
   * Disconnect from the command.
   * 
   * The command is not killed, but SDK stops receiving events from the command.
   * You can reconnect to the command using {@link Commands.connect}.
   */
  disconnect() {
    return __async(this, null, function* () {
      this.handleDisconnect();
    });
  }
  /**
   * Kill the command.
   * It uses `SIGKILL` signal to kill the command.
   * 
   * @returns `true` if the command was killed successfully, `false` if the command was not found.
   */
  kill() {
    return __async(this, null, function* () {
      return yield this.handleKill();
    });
  }
  iterateEvents() {
    return __asyncGenerator(this, null, function* () {
      var _a2;
      try {
        for (var iter = __forAwait(this.events), more, temp, error; more = !(temp = yield new __await(iter.next())).done; more = false) {
          const event = temp.value;
          const e = (_a2 = event == null ? void 0 : event.event) == null ? void 0 : _a2.event;
          let out;
          switch (e == null ? void 0 : e.case) {
            case "data":
              switch (e.value.output.case) {
                case "stdout":
                  out = new TextDecoder().decode(e.value.output.value);
                  this._stdout += out;
                  yield [out, null, null];
                  break;
                case "stderr":
                  out = new TextDecoder().decode(e.value.output.value);
                  this._stderr += out;
                  yield [null, out, null];
                  break;
                case "pty":
                  yield [null, null, e.value.output.value];
                  break;
              }
              break;
            case "end":
              this.result = {
                exitCode: e.value.exitCode,
                error: e.value.error,
                stdout: this.stdout,
                stderr: this.stderr
              };
              break;
          }
        }
      } catch (temp) {
        error = [temp];
      } finally {
        try {
          more && (temp = iter.return) && (yield new __await(temp.call(iter)));
        } finally {
          if (error)
            throw error[0];
        }
      }
    });
  }
  handleEvents() {
    return __async(this, null, function* () {
      var _a2, _b, _c;
      try {
        try {
          for (var iter = __forAwait(this.iterateEvents()), more, temp, error; more = !(temp = yield iter.next()).done; more = false) {
            const [stdout, stderr, pty] = temp.value;
            if (stdout !== null) {
              (_a2 = this.onStdout) == null ? void 0 : _a2.call(this, stdout);
            } else if (stderr !== null) {
              (_b = this.onStderr) == null ? void 0 : _b.call(this, stderr);
            } else if (pty) {
              (_c = this.onPty) == null ? void 0 : _c.call(this, pty);
            }
          }
        } catch (temp) {
          error = [temp];
        } finally {
          try {
            more && (temp = iter.return) && (yield temp.call(iter));
          } finally {
            if (error)
              throw error[0];
          }
        }
      } catch (e) {
        this.iterationError = handleRpcError(e);
      }
    });
  }
};

// src/sandbox/index.ts
import { createConnectTransport } from "@connectrpc/connect-web";

// src/sandbox/commands/index.ts
import {
  Code as Code5,
  ConnectError as ConnectError5,
  createPromiseClient as createPromiseClient3
} from "@connectrpc/connect";

// src/envd/process/process_pb.ts
import { Message as Message2, proto3 as proto32 } from "@bufbuild/protobuf";
var Signal = /* @__PURE__ */ ((Signal2) => {
  Signal2[Signal2["UNSPECIFIED"] = 0] = "UNSPECIFIED";
  Signal2[Signal2["SIGTERM"] = 15] = "SIGTERM";
  Signal2[Signal2["SIGKILL"] = 9] = "SIGKILL";
  return Signal2;
})(Signal || {});
proto32.util.setEnumType(Signal, "process.Signal", [
  { no: 0, name: "SIGNAL_UNSPECIFIED" },
  { no: 15, name: "SIGNAL_SIGTERM" },
  { no: 9, name: "SIGNAL_SIGKILL" }
]);
var _PTY = class _PTY extends Message2 {
  constructor(data) {
    super();
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _PTY().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _PTY().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _PTY().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_PTY, a, b);
  }
};
_PTY.runtime = proto32;
_PTY.typeName = "process.PTY";
_PTY.fields = proto32.util.newFieldList(() => [
  { no: 1, name: "size", kind: "message", T: PTY_Size }
]);
var PTY = _PTY;
var _PTY_Size = class _PTY_Size extends Message2 {
  constructor(data) {
    super();
    /**
     * @generated from field: uint32 cols = 1;
     */
    this.cols = 0;
    /**
     * @generated from field: uint32 rows = 2;
     */
    this.rows = 0;
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _PTY_Size().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _PTY_Size().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _PTY_Size().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_PTY_Size, a, b);
  }
};
_PTY_Size.runtime = proto32;
_PTY_Size.typeName = "process.PTY.Size";
_PTY_Size.fields = proto32.util.newFieldList(() => [
  {
    no: 1,
    name: "cols",
    kind: "scalar",
    T: 13
    /* ScalarType.UINT32 */
  },
  {
    no: 2,
    name: "rows",
    kind: "scalar",
    T: 13
    /* ScalarType.UINT32 */
  }
]);
var PTY_Size = _PTY_Size;
var _ProcessConfig = class _ProcessConfig extends Message2 {
  constructor(data) {
    super();
    /**
     * @generated from field: string cmd = 1;
     */
    this.cmd = "";
    /**
     * @generated from field: repeated string args = 2;
     */
    this.args = [];
    /**
     * @generated from field: map<string, string> envs = 3;
     */
    this.envs = {};
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _ProcessConfig().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _ProcessConfig().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _ProcessConfig().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_ProcessConfig, a, b);
  }
};
_ProcessConfig.runtime = proto32;
_ProcessConfig.typeName = "process.ProcessConfig";
_ProcessConfig.fields = proto32.util.newFieldList(() => [
  {
    no: 1,
    name: "cmd",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  { no: 2, name: "args", kind: "scalar", T: 9, repeated: true },
  { no: 3, name: "envs", kind: "map", K: 9, V: {
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  } },
  { no: 4, name: "cwd", kind: "scalar", T: 9, opt: true }
]);
var ProcessConfig = _ProcessConfig;
var _ListRequest = class _ListRequest extends Message2 {
  constructor(data) {
    super();
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _ListRequest().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _ListRequest().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _ListRequest().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_ListRequest, a, b);
  }
};
_ListRequest.runtime = proto32;
_ListRequest.typeName = "process.ListRequest";
_ListRequest.fields = proto32.util.newFieldList(() => []);
var ListRequest = _ListRequest;
var _ProcessInfo = class _ProcessInfo extends Message2 {
  constructor(data) {
    super();
    /**
     * @generated from field: uint32 pid = 2;
     */
    this.pid = 0;
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _ProcessInfo().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _ProcessInfo().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _ProcessInfo().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_ProcessInfo, a, b);
  }
};
_ProcessInfo.runtime = proto32;
_ProcessInfo.typeName = "process.ProcessInfo";
_ProcessInfo.fields = proto32.util.newFieldList(() => [
  { no: 1, name: "config", kind: "message", T: ProcessConfig },
  {
    no: 2,
    name: "pid",
    kind: "scalar",
    T: 13
    /* ScalarType.UINT32 */
  },
  { no: 3, name: "tag", kind: "scalar", T: 9, opt: true }
]);
var ProcessInfo = _ProcessInfo;
var _ListResponse = class _ListResponse extends Message2 {
  constructor(data) {
    super();
    /**
     * @generated from field: repeated process.ProcessInfo processes = 1;
     */
    this.processes = [];
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _ListResponse().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _ListResponse().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _ListResponse().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_ListResponse, a, b);
  }
};
_ListResponse.runtime = proto32;
_ListResponse.typeName = "process.ListResponse";
_ListResponse.fields = proto32.util.newFieldList(() => [
  { no: 1, name: "processes", kind: "message", T: ProcessInfo, repeated: true }
]);
var ListResponse = _ListResponse;
var _StartRequest = class _StartRequest extends Message2 {
  constructor(data) {
    super();
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _StartRequest().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _StartRequest().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _StartRequest().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_StartRequest, a, b);
  }
};
_StartRequest.runtime = proto32;
_StartRequest.typeName = "process.StartRequest";
_StartRequest.fields = proto32.util.newFieldList(() => [
  { no: 1, name: "process", kind: "message", T: ProcessConfig },
  { no: 2, name: "pty", kind: "message", T: PTY, opt: true },
  { no: 3, name: "tag", kind: "scalar", T: 9, opt: true }
]);
var StartRequest = _StartRequest;
var _UpdateRequest = class _UpdateRequest extends Message2 {
  constructor(data) {
    super();
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _UpdateRequest().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _UpdateRequest().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _UpdateRequest().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_UpdateRequest, a, b);
  }
};
_UpdateRequest.runtime = proto32;
_UpdateRequest.typeName = "process.UpdateRequest";
_UpdateRequest.fields = proto32.util.newFieldList(() => [
  { no: 1, name: "process", kind: "message", T: ProcessSelector },
  { no: 2, name: "pty", kind: "message", T: PTY, opt: true }
]);
var UpdateRequest = _UpdateRequest;
var _UpdateResponse = class _UpdateResponse extends Message2 {
  constructor(data) {
    super();
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _UpdateResponse().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _UpdateResponse().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _UpdateResponse().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_UpdateResponse, a, b);
  }
};
_UpdateResponse.runtime = proto32;
_UpdateResponse.typeName = "process.UpdateResponse";
_UpdateResponse.fields = proto32.util.newFieldList(() => []);
var UpdateResponse = _UpdateResponse;
var _ProcessEvent = class _ProcessEvent extends Message2 {
  constructor(data) {
    super();
    /**
     * @generated from oneof process.ProcessEvent.event
     */
    this.event = { case: void 0 };
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _ProcessEvent().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _ProcessEvent().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _ProcessEvent().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_ProcessEvent, a, b);
  }
};
_ProcessEvent.runtime = proto32;
_ProcessEvent.typeName = "process.ProcessEvent";
_ProcessEvent.fields = proto32.util.newFieldList(() => [
  { no: 1, name: "start", kind: "message", T: ProcessEvent_StartEvent, oneof: "event" },
  { no: 2, name: "data", kind: "message", T: ProcessEvent_DataEvent, oneof: "event" },
  { no: 3, name: "end", kind: "message", T: ProcessEvent_EndEvent, oneof: "event" },
  { no: 4, name: "keepalive", kind: "message", T: ProcessEvent_KeepAlive, oneof: "event" }
]);
var ProcessEvent = _ProcessEvent;
var _ProcessEvent_StartEvent = class _ProcessEvent_StartEvent extends Message2 {
  constructor(data) {
    super();
    /**
     * @generated from field: uint32 pid = 1;
     */
    this.pid = 0;
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _ProcessEvent_StartEvent().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _ProcessEvent_StartEvent().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _ProcessEvent_StartEvent().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_ProcessEvent_StartEvent, a, b);
  }
};
_ProcessEvent_StartEvent.runtime = proto32;
_ProcessEvent_StartEvent.typeName = "process.ProcessEvent.StartEvent";
_ProcessEvent_StartEvent.fields = proto32.util.newFieldList(() => [
  {
    no: 1,
    name: "pid",
    kind: "scalar",
    T: 13
    /* ScalarType.UINT32 */
  }
]);
var ProcessEvent_StartEvent = _ProcessEvent_StartEvent;
var _ProcessEvent_DataEvent = class _ProcessEvent_DataEvent extends Message2 {
  constructor(data) {
    super();
    /**
     * @generated from oneof process.ProcessEvent.DataEvent.output
     */
    this.output = { case: void 0 };
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _ProcessEvent_DataEvent().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _ProcessEvent_DataEvent().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _ProcessEvent_DataEvent().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_ProcessEvent_DataEvent, a, b);
  }
};
_ProcessEvent_DataEvent.runtime = proto32;
_ProcessEvent_DataEvent.typeName = "process.ProcessEvent.DataEvent";
_ProcessEvent_DataEvent.fields = proto32.util.newFieldList(() => [
  { no: 1, name: "stdout", kind: "scalar", T: 12, oneof: "output" },
  { no: 2, name: "stderr", kind: "scalar", T: 12, oneof: "output" },
  { no: 3, name: "pty", kind: "scalar", T: 12, oneof: "output" }
]);
var ProcessEvent_DataEvent = _ProcessEvent_DataEvent;
var _ProcessEvent_EndEvent = class _ProcessEvent_EndEvent extends Message2 {
  constructor(data) {
    super();
    /**
     * @generated from field: sint32 exit_code = 1;
     */
    this.exitCode = 0;
    /**
     * @generated from field: bool exited = 2;
     */
    this.exited = false;
    /**
     * @generated from field: string status = 3;
     */
    this.status = "";
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _ProcessEvent_EndEvent().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _ProcessEvent_EndEvent().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _ProcessEvent_EndEvent().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_ProcessEvent_EndEvent, a, b);
  }
};
_ProcessEvent_EndEvent.runtime = proto32;
_ProcessEvent_EndEvent.typeName = "process.ProcessEvent.EndEvent";
_ProcessEvent_EndEvent.fields = proto32.util.newFieldList(() => [
  {
    no: 1,
    name: "exit_code",
    kind: "scalar",
    T: 17
    /* ScalarType.SINT32 */
  },
  {
    no: 2,
    name: "exited",
    kind: "scalar",
    T: 8
    /* ScalarType.BOOL */
  },
  {
    no: 3,
    name: "status",
    kind: "scalar",
    T: 9
    /* ScalarType.STRING */
  },
  { no: 4, name: "error", kind: "scalar", T: 9, opt: true }
]);
var ProcessEvent_EndEvent = _ProcessEvent_EndEvent;
var _ProcessEvent_KeepAlive = class _ProcessEvent_KeepAlive extends Message2 {
  constructor(data) {
    super();
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _ProcessEvent_KeepAlive().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _ProcessEvent_KeepAlive().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _ProcessEvent_KeepAlive().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_ProcessEvent_KeepAlive, a, b);
  }
};
_ProcessEvent_KeepAlive.runtime = proto32;
_ProcessEvent_KeepAlive.typeName = "process.ProcessEvent.KeepAlive";
_ProcessEvent_KeepAlive.fields = proto32.util.newFieldList(() => []);
var ProcessEvent_KeepAlive = _ProcessEvent_KeepAlive;
var _StartResponse = class _StartResponse extends Message2 {
  constructor(data) {
    super();
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _StartResponse().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _StartResponse().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _StartResponse().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_StartResponse, a, b);
  }
};
_StartResponse.runtime = proto32;
_StartResponse.typeName = "process.StartResponse";
_StartResponse.fields = proto32.util.newFieldList(() => [
  { no: 1, name: "event", kind: "message", T: ProcessEvent }
]);
var StartResponse = _StartResponse;
var _ConnectResponse = class _ConnectResponse extends Message2 {
  constructor(data) {
    super();
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _ConnectResponse().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _ConnectResponse().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _ConnectResponse().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_ConnectResponse, a, b);
  }
};
_ConnectResponse.runtime = proto32;
_ConnectResponse.typeName = "process.ConnectResponse";
_ConnectResponse.fields = proto32.util.newFieldList(() => [
  { no: 1, name: "event", kind: "message", T: ProcessEvent }
]);
var ConnectResponse = _ConnectResponse;
var _SendInputRequest = class _SendInputRequest extends Message2 {
  constructor(data) {
    super();
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _SendInputRequest().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _SendInputRequest().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _SendInputRequest().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_SendInputRequest, a, b);
  }
};
_SendInputRequest.runtime = proto32;
_SendInputRequest.typeName = "process.SendInputRequest";
_SendInputRequest.fields = proto32.util.newFieldList(() => [
  { no: 1, name: "process", kind: "message", T: ProcessSelector },
  { no: 2, name: "input", kind: "message", T: ProcessInput }
]);
var SendInputRequest = _SendInputRequest;
var _SendInputResponse = class _SendInputResponse extends Message2 {
  constructor(data) {
    super();
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _SendInputResponse().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _SendInputResponse().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _SendInputResponse().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_SendInputResponse, a, b);
  }
};
_SendInputResponse.runtime = proto32;
_SendInputResponse.typeName = "process.SendInputResponse";
_SendInputResponse.fields = proto32.util.newFieldList(() => []);
var SendInputResponse = _SendInputResponse;
var _ProcessInput = class _ProcessInput extends Message2 {
  constructor(data) {
    super();
    /**
     * @generated from oneof process.ProcessInput.input
     */
    this.input = { case: void 0 };
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _ProcessInput().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _ProcessInput().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _ProcessInput().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_ProcessInput, a, b);
  }
};
_ProcessInput.runtime = proto32;
_ProcessInput.typeName = "process.ProcessInput";
_ProcessInput.fields = proto32.util.newFieldList(() => [
  { no: 1, name: "stdin", kind: "scalar", T: 12, oneof: "input" },
  { no: 2, name: "pty", kind: "scalar", T: 12, oneof: "input" }
]);
var ProcessInput = _ProcessInput;
var _StreamInputRequest = class _StreamInputRequest extends Message2 {
  constructor(data) {
    super();
    /**
     * @generated from oneof process.StreamInputRequest.event
     */
    this.event = { case: void 0 };
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _StreamInputRequest().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _StreamInputRequest().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _StreamInputRequest().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_StreamInputRequest, a, b);
  }
};
_StreamInputRequest.runtime = proto32;
_StreamInputRequest.typeName = "process.StreamInputRequest";
_StreamInputRequest.fields = proto32.util.newFieldList(() => [
  { no: 1, name: "start", kind: "message", T: StreamInputRequest_StartEvent, oneof: "event" },
  { no: 2, name: "data", kind: "message", T: StreamInputRequest_DataEvent, oneof: "event" },
  { no: 3, name: "keepalive", kind: "message", T: StreamInputRequest_KeepAlive, oneof: "event" }
]);
var StreamInputRequest = _StreamInputRequest;
var _StreamInputRequest_StartEvent = class _StreamInputRequest_StartEvent extends Message2 {
  constructor(data) {
    super();
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _StreamInputRequest_StartEvent().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _StreamInputRequest_StartEvent().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _StreamInputRequest_StartEvent().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_StreamInputRequest_StartEvent, a, b);
  }
};
_StreamInputRequest_StartEvent.runtime = proto32;
_StreamInputRequest_StartEvent.typeName = "process.StreamInputRequest.StartEvent";
_StreamInputRequest_StartEvent.fields = proto32.util.newFieldList(() => [
  { no: 1, name: "process", kind: "message", T: ProcessSelector }
]);
var StreamInputRequest_StartEvent = _StreamInputRequest_StartEvent;
var _StreamInputRequest_DataEvent = class _StreamInputRequest_DataEvent extends Message2 {
  constructor(data) {
    super();
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _StreamInputRequest_DataEvent().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _StreamInputRequest_DataEvent().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _StreamInputRequest_DataEvent().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_StreamInputRequest_DataEvent, a, b);
  }
};
_StreamInputRequest_DataEvent.runtime = proto32;
_StreamInputRequest_DataEvent.typeName = "process.StreamInputRequest.DataEvent";
_StreamInputRequest_DataEvent.fields = proto32.util.newFieldList(() => [
  { no: 2, name: "input", kind: "message", T: ProcessInput }
]);
var StreamInputRequest_DataEvent = _StreamInputRequest_DataEvent;
var _StreamInputRequest_KeepAlive = class _StreamInputRequest_KeepAlive extends Message2 {
  constructor(data) {
    super();
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _StreamInputRequest_KeepAlive().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _StreamInputRequest_KeepAlive().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _StreamInputRequest_KeepAlive().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_StreamInputRequest_KeepAlive, a, b);
  }
};
_StreamInputRequest_KeepAlive.runtime = proto32;
_StreamInputRequest_KeepAlive.typeName = "process.StreamInputRequest.KeepAlive";
_StreamInputRequest_KeepAlive.fields = proto32.util.newFieldList(() => []);
var StreamInputRequest_KeepAlive = _StreamInputRequest_KeepAlive;
var _StreamInputResponse = class _StreamInputResponse extends Message2 {
  constructor(data) {
    super();
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _StreamInputResponse().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _StreamInputResponse().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _StreamInputResponse().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_StreamInputResponse, a, b);
  }
};
_StreamInputResponse.runtime = proto32;
_StreamInputResponse.typeName = "process.StreamInputResponse";
_StreamInputResponse.fields = proto32.util.newFieldList(() => []);
var StreamInputResponse = _StreamInputResponse;
var _SendSignalRequest = class _SendSignalRequest extends Message2 {
  constructor(data) {
    super();
    /**
     * @generated from field: process.Signal signal = 2;
     */
    this.signal = 0 /* UNSPECIFIED */;
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _SendSignalRequest().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _SendSignalRequest().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _SendSignalRequest().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_SendSignalRequest, a, b);
  }
};
_SendSignalRequest.runtime = proto32;
_SendSignalRequest.typeName = "process.SendSignalRequest";
_SendSignalRequest.fields = proto32.util.newFieldList(() => [
  { no: 1, name: "process", kind: "message", T: ProcessSelector },
  { no: 2, name: "signal", kind: "enum", T: proto32.getEnumType(Signal) }
]);
var SendSignalRequest = _SendSignalRequest;
var _SendSignalResponse = class _SendSignalResponse extends Message2 {
  constructor(data) {
    super();
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _SendSignalResponse().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _SendSignalResponse().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _SendSignalResponse().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_SendSignalResponse, a, b);
  }
};
_SendSignalResponse.runtime = proto32;
_SendSignalResponse.typeName = "process.SendSignalResponse";
_SendSignalResponse.fields = proto32.util.newFieldList(() => []);
var SendSignalResponse = _SendSignalResponse;
var _ConnectRequest = class _ConnectRequest extends Message2 {
  constructor(data) {
    super();
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _ConnectRequest().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _ConnectRequest().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _ConnectRequest().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_ConnectRequest, a, b);
  }
};
_ConnectRequest.runtime = proto32;
_ConnectRequest.typeName = "process.ConnectRequest";
_ConnectRequest.fields = proto32.util.newFieldList(() => [
  { no: 1, name: "process", kind: "message", T: ProcessSelector }
]);
var ConnectRequest = _ConnectRequest;
var _ProcessSelector = class _ProcessSelector extends Message2 {
  constructor(data) {
    super();
    /**
     * @generated from oneof process.ProcessSelector.selector
     */
    this.selector = { case: void 0 };
    proto32.util.initPartial(data, this);
  }
  static fromBinary(bytes, options) {
    return new _ProcessSelector().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new _ProcessSelector().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new _ProcessSelector().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return proto32.util.equals(_ProcessSelector, a, b);
  }
};
_ProcessSelector.runtime = proto32;
_ProcessSelector.typeName = "process.ProcessSelector";
_ProcessSelector.fields = proto32.util.newFieldList(() => [
  { no: 1, name: "pid", kind: "scalar", T: 13, oneof: "selector" },
  { no: 2, name: "tag", kind: "scalar", T: 9, oneof: "selector" }
]);
var ProcessSelector = _ProcessSelector;

// src/envd/process/process_connect.ts
import { MethodKind as MethodKind2 } from "@bufbuild/protobuf";
var Process = {
  typeName: "process.Process",
  methods: {
    /**
     * @generated from rpc process.Process.List
     */
    list: {
      name: "List",
      I: ListRequest,
      O: ListResponse,
      kind: MethodKind2.Unary
    },
    /**
     * @generated from rpc process.Process.Connect
     */
    connect: {
      name: "Connect",
      I: ConnectRequest,
      O: ConnectResponse,
      kind: MethodKind2.ServerStreaming
    },
    /**
     * @generated from rpc process.Process.Start
     */
    start: {
      name: "Start",
      I: StartRequest,
      O: StartResponse,
      kind: MethodKind2.ServerStreaming
    },
    /**
     * @generated from rpc process.Process.Update
     */
    update: {
      name: "Update",
      I: UpdateRequest,
      O: UpdateResponse,
      kind: MethodKind2.Unary
    },
    /**
     * Client input stream ensures ordering of messages
     *
     * @generated from rpc process.Process.StreamInput
     */
    streamInput: {
      name: "StreamInput",
      I: StreamInputRequest,
      O: StreamInputResponse,
      kind: MethodKind2.ClientStreaming
    },
    /**
     * @generated from rpc process.Process.SendInput
     */
    sendInput: {
      name: "SendInput",
      I: SendInputRequest,
      O: SendInputResponse,
      kind: MethodKind2.Unary
    },
    /**
     * @generated from rpc process.Process.SendSignal
     */
    sendSignal: {
      name: "SendSignal",
      I: SendSignalRequest,
      O: SendSignalResponse,
      kind: MethodKind2.Unary
    }
  }
};

// src/sandbox/commands/pty.ts
import {
  Code as Code4,
  ConnectError as ConnectError4,
  createPromiseClient as createPromiseClient2
} from "@connectrpc/connect";
var Pty = class {
  constructor(transport, connectionConfig) {
    this.transport = transport;
    this.connectionConfig = connectionConfig;
    this.rpc = createPromiseClient2(Process, this.transport);
  }
  /**
   * Create a new PTY (pseudo-terminal).
   *
   * @param opts options for creating the PTY.
   * 
   * @returns handle to interact with the PTY.
   */
  create(opts) {
    return __async(this, null, function* () {
      var _a2, _b, _c;
      const requestTimeoutMs = (_a2 = opts == null ? void 0 : opts.requestTimeoutMs) != null ? _a2 : this.connectionConfig.requestTimeoutMs;
      const envs = (_b = opts == null ? void 0 : opts.envs) != null ? _b : {};
      envs.TERM = "xterm-256color";
      const controller = new AbortController();
      const reqTimeout = setTimeout(() => {
        controller.abort();
      }, requestTimeoutMs);
      const events = this.rpc.start(
        {
          process: {
            cmd: "/bin/bash",
            args: ["-i", "-l"],
            envs,
            cwd: opts == null ? void 0 : opts.cwd
          },
          pty: {
            size: {
              cols: opts.cols,
              rows: opts.rows
            }
          }
        },
        {
          headers: __spreadProps(__spreadValues({}, authenticationHeader(opts == null ? void 0 : opts.user)), {
            [KEEPALIVE_PING_HEADER]: KEEPALIVE_PING_INTERVAL_SEC.toString()
          }),
          signal: controller.signal,
          timeoutMs: (_c = opts == null ? void 0 : opts.timeoutMs) != null ? _c : 6e4
        }
      );
      try {
        const pid = yield handleProcessStartEvent(events);
        clearTimeout(reqTimeout);
        return new CommandHandle(
          pid,
          () => controller.abort(),
          () => this.kill(pid),
          events,
          void 0,
          void 0,
          opts.onData
        );
      } catch (err) {
        throw handleRpcError(err);
      }
    });
  }
  /**
   * Send input to a PTY.
   *
   * @param pid process ID of the PTY.
   * @param data input data to send to the PTY.
   * @param opts connection options.
   */
  sendInput(pid, data, opts) {
    return __async(this, null, function* () {
      try {
        yield this.rpc.sendInput(
          {
            input: {
              input: {
                case: "pty",
                value: data
              }
            },
            process: {
              selector: {
                case: "pid",
                value: pid
              }
            }
          },
          {
            signal: this.connectionConfig.getSignal(opts == null ? void 0 : opts.requestTimeoutMs)
          }
        );
      } catch (err) {
        throw handleRpcError(err);
      }
    });
  }
  /**
   * Resize PTY.
   * Call this when the terminal window is resized and the number of columns and rows has changed.
   *
   * @param pid process ID of the PTY.
   * @param size new size of the PTY.
   * @param opts connection options.
   */
  resize(pid, size, opts) {
    return __async(this, null, function* () {
      try {
        yield this.rpc.update(
          {
            process: {
              selector: {
                case: "pid",
                value: pid
              }
            },
            pty: {
              size
            }
          },
          {
            signal: this.connectionConfig.getSignal(opts == null ? void 0 : opts.requestTimeoutMs)
          }
        );
      } catch (err) {
        throw handleRpcError(err);
      }
    });
  }
  /**
   * Kill a running PTY specified by process ID.
   * It uses `SIGKILL` signal to kill the PTY.
   *
   * @param pid process ID of the PTY.
   * @param opts connection options.
   * 
   * @returns `true` if the PTY was killed, `false` if the PTY was not found.
   */
  kill(pid, opts) {
    return __async(this, null, function* () {
      try {
        yield this.rpc.sendSignal(
          {
            process: {
              selector: {
                case: "pid",
                value: pid
              }
            },
            signal: 9 /* SIGKILL */
          },
          {
            signal: this.connectionConfig.getSignal(opts == null ? void 0 : opts.requestTimeoutMs)
          }
        );
        return true;
      } catch (err) {
        if (err instanceof ConnectError4) {
          if (err.code === Code4.NotFound) {
            return false;
          }
        }
        throw handleRpcError(err);
      }
    });
  }
};

// src/sandbox/commands/index.ts
var Commands = class {
  // 60 seconds
  constructor(transport, connectionConfig) {
    this.connectionConfig = connectionConfig;
    this.defaultProcessConnectionTimeout = 6e4;
    this.rpc = createPromiseClient3(Process, transport);
  }
  /**
   * List all running commands and PTY sessions.
   *
   * @param opts connection options.
   * 
   * @returns list of running commands and PTY sessions.
   */
  list(opts) {
    return __async(this, null, function* () {
      try {
        const res = yield this.rpc.list(
          {},
          {
            signal: this.connectionConfig.getSignal(opts == null ? void 0 : opts.requestTimeoutMs)
          }
        );
        return res.processes.map((p) => __spreadValues(__spreadProps(__spreadValues({
          pid: p.pid
        }, p.tag && { tag: p.tag }), {
          args: p.config.args,
          envs: p.config.envs,
          cmd: p.config.cmd
        }), p.config.cwd && { cwd: p.config.cwd }));
      } catch (err) {
        throw handleRpcError(err);
      }
    });
  }
  /**
   * Send data to command stdin.
   *
   * @param pid process ID of the command. You can get the list of running commands using {@link Commands.list}.
   * @param data data to send to the command.
   * @param opts connection options.
   */
  sendStdin(pid, data, opts) {
    return __async(this, null, function* () {
      try {
        yield this.rpc.sendInput(
          {
            process: {
              selector: {
                case: "pid",
                value: pid
              }
            },
            input: {
              input: {
                case: "stdin",
                value: new TextEncoder().encode(data)
              }
            }
          },
          {
            signal: this.connectionConfig.getSignal(opts == null ? void 0 : opts.requestTimeoutMs)
          }
        );
      } catch (err) {
        throw handleRpcError(err);
      }
    });
  }
  /**
   * Kill a running command specified by its process ID.
   * It uses `SIGKILL` signal to kill the command.
   *
   * @param pid process ID of the command. You can get the list of running commands using {@link Commands.list}.
   * @param opts connection options.
   * 
   * @returns `true` if the command was killed, `false` if the command was not found.
   */
  kill(pid, opts) {
    return __async(this, null, function* () {
      try {
        yield this.rpc.sendSignal(
          {
            process: {
              selector: {
                case: "pid",
                value: pid
              }
            },
            signal: 9 /* SIGKILL */
          },
          {
            signal: this.connectionConfig.getSignal(opts == null ? void 0 : opts.requestTimeoutMs)
          }
        );
        return true;
      } catch (err) {
        if (err instanceof ConnectError5) {
          if (err.code === Code5.NotFound) {
            return false;
          }
        }
        throw handleRpcError(err);
      }
    });
  }
  /**
   * Connect to a running command.
   * You can use {@link CommandHandle.wait} to wait for the command to finish and get execution results.
   *
   * @param pid process ID of the command to connect to. You can get the list of running commands using {@link Commands.list}.
   * @param opts connection options.
   * 
   * @returns `CommandHandle` handle to interact with the running command.
   */
  connect(pid, opts) {
    return __async(this, null, function* () {
      var _a2, _b;
      const requestTimeoutMs = (_a2 = opts == null ? void 0 : opts.requestTimeoutMs) != null ? _a2 : this.connectionConfig.requestTimeoutMs;
      const controller = new AbortController();
      const reqTimeout = requestTimeoutMs ? setTimeout(() => {
        controller.abort();
      }, requestTimeoutMs) : void 0;
      const events = this.rpc.connect(
        {
          process: {
            selector: {
              case: "pid",
              value: pid
            }
          }
        },
        {
          signal: controller.signal,
          headers: {
            [KEEPALIVE_PING_HEADER]: KEEPALIVE_PING_INTERVAL_SEC.toString()
          },
          timeoutMs: (_b = opts == null ? void 0 : opts.timeoutMs) != null ? _b : this.defaultProcessConnectionTimeout
        }
      );
      try {
        const pid2 = yield handleProcessStartEvent(events);
        clearTimeout(reqTimeout);
        return new CommandHandle(
          pid2,
          () => controller.abort(),
          () => this.kill(pid2),
          events,
          opts == null ? void 0 : opts.onStdout,
          opts == null ? void 0 : opts.onStderr,
          void 0
        );
      } catch (err) {
        throw handleRpcError(err);
      }
    });
  }
  run(cmd, opts) {
    return __async(this, null, function* () {
      const proc = yield this.start(cmd, opts);
      return (opts == null ? void 0 : opts.background) ? proc : proc.wait();
    });
  }
  start(cmd, opts) {
    return __async(this, null, function* () {
      var _a2, _b;
      const requestTimeoutMs = (_a2 = opts == null ? void 0 : opts.requestTimeoutMs) != null ? _a2 : this.connectionConfig.requestTimeoutMs;
      const controller = new AbortController();
      const reqTimeout = requestTimeoutMs ? setTimeout(() => {
        controller.abort();
      }, requestTimeoutMs) : void 0;
      const events = this.rpc.start(
        {
          process: {
            cmd: "/bin/bash",
            cwd: opts == null ? void 0 : opts.cwd,
            envs: opts == null ? void 0 : opts.envs,
            args: ["-l", "-c", cmd]
          }
        },
        {
          headers: __spreadProps(__spreadValues({}, authenticationHeader(opts == null ? void 0 : opts.user)), {
            [KEEPALIVE_PING_HEADER]: KEEPALIVE_PING_INTERVAL_SEC.toString()
          }),
          signal: controller.signal,
          timeoutMs: (_b = opts == null ? void 0 : opts.timeoutMs) != null ? _b : this.defaultProcessConnectionTimeout
        }
      );
      try {
        const pid = yield handleProcessStartEvent(events);
        clearTimeout(reqTimeout);
        return new CommandHandle(
          pid,
          () => controller.abort(),
          () => this.kill(pid),
          events,
          opts == null ? void 0 : opts.onStdout,
          opts == null ? void 0 : opts.onStderr,
          void 0
        );
      } catch (err) {
        throw handleRpcError(err);
      }
    });
  }
};

// src/sandbox/sandboxApi.ts
import { compareVersions } from "compare-versions";
var SandboxApi = class {
  constructor() {
  }
  /**
   * Kill the sandbox specified by sandbox ID.
   *
   * @param sandboxId sandbox ID.
   * @param opts connection options.
   *
   * @returns `true` if the sandbox was found and killed, `false` otherwise.
   */
  static kill(sandboxId, opts) {
    return __async(this, null, function* () {
      var _a2;
      const config = new ConnectionConfig(opts);
      const client = new ApiClient(config);
      const res = yield client.api.DELETE("/sandboxes/{sandboxID}", {
        params: {
          path: {
            sandboxID: sandboxId
          }
        },
        signal: config.getSignal(opts == null ? void 0 : opts.requestTimeoutMs)
      });
      if (((_a2 = res.error) == null ? void 0 : _a2.code) === 404) {
        return false;
      }
      const err = handleApiError(res);
      if (err) {
        throw err;
      }
      return true;
    });
  }
  /**
   * List all running sandboxes.
   *
   * @param opts connection options.
   *
   * @returns list of running sandboxes.
   */
  static list(opts) {
    return __async(this, null, function* () {
      var _a2, _b;
      const config = new ConnectionConfig(opts);
      const client = new ApiClient(config);
      const res = yield client.api.GET("/sandboxes", {
        signal: config.getSignal(opts == null ? void 0 : opts.requestTimeoutMs)
      });
      const err = handleApiError(res);
      if (err) {
        throw err;
      }
      return (_b = (_a2 = res.data) == null ? void 0 : _a2.map((sandbox) => {
        var _a3;
        return __spreadProps(__spreadValues({
          sandboxId: this.getSandboxId({
            sandboxId: sandbox.sandboxID,
            clientId: sandbox.clientID
          }),
          templateId: sandbox.templateID
        }, sandbox.alias && { name: sandbox.alias }), {
          metadata: (_a3 = sandbox.metadata) != null ? _a3 : {},
          startedAt: new Date(sandbox.startedAt)
        });
      })) != null ? _b : [];
    });
  }
  /**
   * Set the timeout of the specified sandbox.
   * After the timeout expires the sandbox will be automatically killed.
   *
   * This method can extend or reduce the sandbox timeout set when creating the sandbox or from the last call to {@link Sandbox.setTimeout}.
   *
   * Maximum time a sandbox can be kept alive is 24 hours (86_400_000 milliseconds) for Pro users and 1 hour (3_600_000 milliseconds) for Hobby users.
   *
   * @param sandboxId sandbox ID.
   * @param timeoutMs timeout in **milliseconds**.
   * @param opts connection options.
   */
  static setTimeout(sandboxId, timeoutMs, opts) {
    return __async(this, null, function* () {
      const config = new ConnectionConfig(opts);
      const client = new ApiClient(config);
      const res = yield client.api.POST("/sandboxes/{sandboxID}/timeout", {
        params: {
          path: {
            sandboxID: sandboxId
          }
        },
        body: {
          timeout: this.timeoutToSeconds(timeoutMs)
        },
        signal: config.getSignal(opts == null ? void 0 : opts.requestTimeoutMs)
      });
      const err = handleApiError(res);
      if (err) {
        throw err;
      }
    });
  }
  static createSandbox(template, timeoutMs, opts) {
    return __async(this, null, function* () {
      const config = new ConnectionConfig(opts);
      const client = new ApiClient(config);
      const res = yield client.api.POST("/sandboxes", {
        body: {
          templateID: template,
          metadata: opts == null ? void 0 : opts.metadata,
          envVars: opts == null ? void 0 : opts.envs,
          timeout: this.timeoutToSeconds(timeoutMs)
        },
        signal: config.getSignal(opts == null ? void 0 : opts.requestTimeoutMs)
      });
      const err = handleApiError(res);
      if (err) {
        throw err;
      }
      if (compareVersions(res.data.envdVersion, "0.1.0") < 0) {
        yield this.kill(
          this.getSandboxId({
            sandboxId: res.data.sandboxID,
            clientId: res.data.clientID
          }),
          opts
        );
        throw new TemplateError(
          "You need to update the template to use the new SDK. You can do this by running `e2b template build` in the directory with the template."
        );
      }
      return this.getSandboxId({
        sandboxId: res.data.sandboxID,
        clientId: res.data.clientID
      });
    });
  }
  static timeoutToSeconds(timeout) {
    return Math.ceil(timeout / 1e3);
  }
  static getSandboxId({
    sandboxId,
    clientId
  }) {
    return `${sandboxId}-${clientId}`;
  }
};

// src/sandbox/index.ts
var _Sandbox = class _Sandbox extends SandboxApi {
  /**
   * Use {@link Sandbox.create} to create a new Sandbox instead.
   *
   * @hidden
   * @hide
   * @internal
   * @access protected
   */
  constructor(opts) {
    super();
    this.envdPort = 49983;
    this.sandboxId = opts.sandboxId;
    this.connectionConfig = new ConnectionConfig(opts);
    this.envdApiUrl = `${this.connectionConfig.debug ? "http" : "https"}://${this.getHost(this.envdPort)}`;
    const rpcTransport = createConnectTransport({
      baseUrl: this.envdApiUrl,
      useBinaryFormat: false,
      interceptors: (opts == null ? void 0 : opts.logger) ? [createRpcLogger(opts.logger)] : void 0
    });
    this.envdApi = new EnvdApiClient({
      apiUrl: this.envdApiUrl,
      logger: opts == null ? void 0 : opts.logger
    });
    this.files = new Filesystem2(
      rpcTransport,
      this.envdApi,
      this.connectionConfig
    );
    this.commands = new Commands(rpcTransport, this.connectionConfig);
    this.pty = new Pty(rpcTransport, this.connectionConfig);
  }
  static create(templateOrOpts, opts) {
    return __async(this, null, function* () {
      var _a2;
      const { template, sandboxOpts } = typeof templateOrOpts === "string" ? { template: templateOrOpts, sandboxOpts: opts } : { template: this.defaultTemplate, sandboxOpts: templateOrOpts };
      const config = new ConnectionConfig(sandboxOpts);
      const sandboxId = config.debug ? "debug_sandbox_id" : yield this.createSandbox(
        template,
        (_a2 = sandboxOpts == null ? void 0 : sandboxOpts.timeoutMs) != null ? _a2 : this.defaultSandboxTimeoutMs,
        sandboxOpts
      );
      const sbx = new this(__spreadValues({ sandboxId }, config));
      return sbx;
    });
  }
  /**
   * Connect to an existing sandbox.
   * With sandbox ID you can connect to the same sandbox from different places or environments (serverless functions, etc).
   * 
   * @param sandboxId sandbox ID.
   * @param opts connection options.
   * 
   * @returns sandbox instance for the existing sandbox.
   *
   * @example
   * ```ts
   * const sandbox = await Sandbox.create()
   * const sandboxId = sandbox.sandboxId
   *
   * // Connect to the same sandbox.
   * const sameSandbox = await Sandbox.connect(sandboxId)
   * ```
   */
  static connect(sandboxId, opts) {
    return __async(this, null, function* () {
      const config = new ConnectionConfig(opts);
      const sbx = new this(__spreadValues({ sandboxId }, config));
      return sbx;
    });
  }
  /**
   * Get the host address for the specified sandbox port.
   * You can then use this address to connect to the sandbox port from outside the sandbox via HTTP or WebSocket.
   *
   * @param port number of the port in the sandbox.
   *
   * @returns host address of the sandbox port.
   * 
   * @example
   * ```ts
   * const sandbox = await Sandbox.create()
   * // Start an HTTP server
   * await sandbox.commands.exec('python3 -m http.server 3000')
   * // Get the hostname of the HTTP server
   * const serverURL = sandbox.getHost(3000)
   * ``
   */
  getHost(port) {
    if (this.connectionConfig.debug) {
      return `localhost:${port}`;
    }
    return `${port}-${this.sandboxId}.${this.connectionConfig.domain}`;
  }
  /**
   * Check if the sandbox is running.
   *
   * @returns `true` if the sandbox is running, `false` otherwise.
   * 
   * @example
   * ```ts
   * const sandbox = await Sandbox.create()
   * await sandbox.isRunning() // Returns true
   *
   * await sandbox.kill()
   * await sandbox.isRunning() // Returns false
   * ```
   */
  isRunning(opts) {
    return __async(this, null, function* () {
      const signal = this.connectionConfig.getSignal(opts == null ? void 0 : opts.requestTimeoutMs);
      const res = yield this.envdApi.api.GET("/health", {
        signal
      });
      if (res.response.status == 502) {
        return false;
      }
      const err = yield handleEnvdApiError(res);
      if (err) {
        throw err;
      }
      return true;
    });
  }
  /**
   * Set the timeout of the sandbox.
   * After the timeout expires the sandbox will be automatically killed.
   * 
   * This method can extend or reduce the sandbox timeout set when creating the sandbox or from the last call to `.setTimeout`.
   * Maximum time a sandbox can be kept alive is 24 hours (86_400_000 milliseconds) for Pro users and 1 hour (3_600_000 milliseconds) for Hobby users.
   *
   * @param timeoutMs timeout in **milliseconds**.
   * @param opts connection options.
   */
  setTimeout(timeoutMs, opts) {
    return __async(this, null, function* () {
      if (this.connectionConfig.debug) {
        return;
      }
      yield _Sandbox.setTimeout(this.sandboxId, timeoutMs, __spreadValues(__spreadValues({}, this.connectionConfig), opts));
    });
  }
  /**
   * Kill the sandbox.
   *
   * @param opts connection options.
   */
  kill(opts) {
    return __async(this, null, function* () {
      if (this.connectionConfig.debug) {
        return;
      }
      yield _Sandbox.kill(this.sandboxId, __spreadValues(__spreadValues({}, this.connectionConfig), opts));
    });
  }
  /**
   * Get the URL to upload a file to the sandbox.
   * 
   * You have to send a POST request to this URL with the file as multipart/form-data.
   * 
   * @param path the directory where to upload the file, defaults to user's home directory.
   * 
   * @returns URL for uploading file.
   */
  uploadUrl(path) {
    return this.fileUrl(path);
  }
  /**
   * Get the URL to download a file from the sandbox.
   *
   * @param path path to the file to download.
   * 
   * @returns URL for downloading file.
   */
  downloadUrl(path) {
    return this.fileUrl(path);
  }
  fileUrl(path) {
    const url = new URL("/files", this.envdApiUrl);
    url.searchParams.set("username", defaultUsername);
    if (path) {
      url.searchParams.set("path", path);
    }
    return url.toString();
  }
};
_Sandbox.defaultTemplate = "base";
_Sandbox.defaultSandboxTimeoutMs = 3e5;
var Sandbox = _Sandbox;

// src/index.ts
var src_default = Sandbox;
export {
  ApiClient,
  AuthenticationError,
  CommandExitError,
  ConnectionConfig,
  FileType2 as FileType,
  FilesystemEventType,
  InvalidArgumentError,
  NotEnoughSpaceError,
  NotFoundError,
  Sandbox,
  SandboxError,
  TemplateError,
  TimeoutError,
  src_default as default
};
